### PR TITLE
Identity V3: Add credentials package (#1457)

### PR DESCRIPTION
**Pull request by himanshuvar**
* Identity V3: Add credentials package

For #1457

Adding the support for Keystone Credential package:

https://github.com/openstack/keystone/tree/master/keystone/credential

### CODE REVIEW

**Comment by kayrus** [comment]
recheck

File: openstack/identity/v3/credentials/requests.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+type CreateOptsBuilder interface {
+	ToCredentialCreateMap() (map[string]interface{}, error)
+}
+
+// CreateOpts provides options used to create a credential.
+type CreateOpts struct {
+	// Serialized blob containing the credentials
+	Blob string `json:"blob" required:"true"`
+	// ID of the project.
+	ProjectID string `json:"project_id" required:"true"`
```
Comment (kayrus) [inline openstack/identity/v3/credentials/requests.go]: `project_id` is not required, when the type is not `ec2`. https://github.com/openstack/keystone/blob/00663f93b7e9513602d73dcd0aba7ad1cc1feefc/keystone/credential/schema.py#L44

File: openstack/identity/v3/credentials/requests.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+type UpdateOptsBuilder interface {
+	ToCredentialsUpdateMap() (map[string]interface{}, error)
+}
+
+// UpdateOpts represents parameters to update a credential.
+type UpdateOpts struct {
+	// Serialized blob containing the credentials.
+	Blob string `json:"blob,omitempty"`
+	// ID of the project.
+	ProjectID string `json:"project_id" required:"true"`
```
Comment (kayrus) [inline openstack/identity/v3/credentials/requests.go]: `project_id` is not required, when the type is not `ec2`. https://github.com/openstack/keystone/blob/00663f93b7e9513602d73dcd0aba7ad1cc1feefc/keystone/credential/schema.py#L44

File: openstack/identity/v3/credentials/requests.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+// ListOptsBuilder allows extensions to add additional parameters to
+// the List request
+type ListOptsBuilder interface {
+	ToCredentialListQuery() (string, error)
+}
+
+// ListOpts provides options to filter the List results.
+type ListOpts struct {
+	// UserID filters the response by a credential user_id
+	UserID string `json:"user_id"`
```
[the remark this answers is not in the document]
Comment (kayrus) [reply openstack/identity/v3/credentials/requests.go]: In addition, looks like you're copying the base structure of the package from the `application_credential`. I made a type there, there should be:

```go
UserID string `q:"user_id"`
```

I'm not sure whether it is mandatory. @jtopjian please advice.

[the remark this answers is not in the document]
Comment (jtopjian) [reply openstack/identity/v3/credentials/requests.go]: Yeah, the `q` tags are mandatory with query strings: https://github.com/gophercloud/gophercloud/blob/master/params.go#L341

File: openstack/identity/v3/credentials/requests.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+type UpdateOptsBuilder interface {
+	ToCredentialsUpdateMap() (map[string]interface{}, error)
+}
+
+// UpdateOpts represents parameters to update a credential.
+type UpdateOpts struct {
+	// Serialized blob containing the credentials.
+	Blob string `json:"blob,omitempty"`
+	// ID of the project.
+	ProjectID string `json:"project_id"`
```
Comment (kayrus) [inline openstack/identity/v3/credentials/requests.go]: Please add `omitempty` tag

File: openstack/identity/v3/credentials/requests.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+type CreateOptsBuilder interface {
+	ToCredentialCreateMap() (map[string]interface{}, error)
+}
+
+// CreateOpts provides options used to create a credential.
+type CreateOpts struct {
+	// Serialized blob containing the credentials
+	Blob string `json:"blob" required:"true"`
+	// ID of the project.
+	ProjectID string `json:"project_id"`
```
Comment (kayrus) [inline openstack/identity/v3/credentials/requests.go]: Please add `omitempty` tag

File: acceptance/openstack/identity/v3/credentials_test.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+func TestCredentialsCRUD(t *testing.T) {
+	clients.RequireAdmin(t)
+
+	client, err := clients.NewIdentityV3Client()
+	th.AssertNoErr(t, err)
+
+	ao, err := openstack.AuthOptionsFromEnv()
+	th.AssertNoErr(t, err)
+
+	authOptions := tokens.AuthOptions{
```
Comment (kayrus) [inline acceptance/openstack/identity/v3/credentials_test.go]: `tokens.AuthOptions` is redundant here, however it is required for application credential to get the user ID.

File: openstack/identity/v3/credentials/requests.go (hunk 4 above)
Comment (himanshuvar) [reply openstack/identity/v3/credentials/requests.go]: Done.
Thank you for the suggestion.

File: openstack/identity/v3/credentials/requests.go (hunk 5 above)
Comment (himanshuvar) [reply openstack/identity/v3/credentials/requests.go]: Done.
Thank you for the suggestion.

File: openstack/identity/v3/credentials/requests.go [hunk 7]
```diff
@@ -57,7 +57,7 @@ type CreateOpts struct {
 	// Serialized blob containing the credentials
 	Blob string `json:"blob" required:"true"`
 	// ID of the project.
-	ProjectID string `json:"project_id"`
+	ProjectID string `json:"project_id",omitempty`
```
Comment (kayrus) [inline openstack/identity/v3/credentials/requests.go]: typo here, change to:

```
ProjectID string `json:"project_id,omitempty"`
```

Comment (himanshuvar) [reply openstack/identity/v3/credentials/requests.go]: Thank you

**Comment by jtopjian** [comment]
@himanshuvar Thank you for working on this. @kayrus Thank you for looking it over, too!

Would it be better to have `Blob` as `map[string]interface{}`? 

I think it should be possible to convert it to a marshalled string in `ToCredentialCreateMap` and unmarshall it in a `Credential` `UnmarshalJSON` method without too much difficulty.

Thoughts?

**Comment by kayrus** [comment]
@jtopjian , unfortunately no, since blob can handle also a cert, or any unstructured string data.

**Comment by jtopjian** [comment]
Ah, understood. Thanks!

**Comment by jtopjian** [comment]
@himanshuvar Is this ready for review/merge?

**Comment by himanshuvar** [comment]
> eview/merge?
Yes

**Pull request merged**